verifier/verifier.go

```diff
@@ -0,0 +1,133 @@
+
+	// The name being checked, e.g. "www.censys.io".
+	Name string
+
+	// Whitelisted is true if the certificate is whitelisted by a given verifier
+	// as trusted, e.g. when Apple whitelisted subset of StartCom certs by SPKI
+	// hash in the wake of the WoSign incidents surrounding misissuance.
+	Whitelisted bool
+
+	// Blacklisted is true if the certificate is blacklisted by a given verifier
+	// as untrusted, e.g. Cloudflare certificates valid at the time of Heartbleed
+	// disclosure in Chrome.
+	Blacklisted bool
+
+	// Revoked is true if the certificate has been revoked and is listed in the
+	// revocation set that is part of the Verifier (e.g. in OneCRL).
+	Revoked bool
+
+	// ValiditionError will be non-nil when there was some sort of error during
+	// validation not involving a name mismatch, e.g. if a chain could not be
+	// built.
+	ValidationError error
+
+	// NameError will be non-nil when there was a mismatch between the name on the
+	// certificate and the name being verified against.
+	NameError error
+
+	// Current is a list of validated certificate chains that are valid at
+	// ValidationTime, starting at the certificate being verified, and ending at a
+	// certificate in the root store.
+	Current [][]*x509.Certificate
+
+	// Expired is a list of certificate chains that were valid at some point,
+	// but not at ValidationTime.
+	Expired [][]*x509.Certificate
+
+	// Never is a list of certificate chains that could never be valid due to
+	// date-related issues, but are otherwise valid.
+	Never [][]*x509.Certificate
+}
+
+// MatchesDomain returns true if NameError == nil and Name != "".
+func (res *VerificationResult) MatchesDomain() bool {
+	return res.NameError == nil && res.Name != ""
+}
+
+// VerifyProcedure is an interface to implement additional browser specific logic at
+// the start and end of verification.
+type VerifyProcedure interface {
+	// TODO
+}
+
+// VerificationOptions contains settings for Verifier.Verify().
+// VerificationOptions should be safely copyable.
+type VerificationOptions struct {
+	VerifyTime time.Time
+	Name       string
+}
+
+func (opt *VerificationOptions) clean() {
+	if opt.VerifyTime.IsZero() {
+		opt.VerifyTime = time.Now()
+	}
+}
+
+// A Verifier represents a context for verifying certificates.
+type Verifier struct {
+	Roots           *x509.CertPool
+	Intermediates   *x509.CertPool
+	VerifyProcedure VerifyProcedure
+}
+
+func (v *Verifier) convertOptions(opt *VerificationOptions) (out x509.VerifyOptions) {
+	opt.clean()
+	out.CurrentTime = opt.VerifyTime
+	out.Roots = v.Roots
+	out.Intermediates = v.Intermediates
+	out.DNSName = opt.Name
+	return
+}
+
+// Verify checks if c chains back to a certificate in Roots, possibly via a
+// certificate in Intermediates, and returns all such chains. It additional
+// checks if the Name in the VerificationOptions matches the name on the
+// certificate. Finally, it checks to see if c is blacklisted or whitelisted by
+// the Verifier.
```

Comment:
I believe that's checked in `x509.Verify`. I should remove the comment about blacklisted vs whitelisted though.